projects/ADuCM3029_demo_cn0414/src/adc_update_timer.c

```diff
@@ -0,0 +1,271 @@
+* INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+* LIMITED TO, INTELLECTUAL PROPERTY RIGHTS, PROCUREMENT OF SUBSTITUTE GOODS OR
+* SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+* CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+* OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+* OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/******************************************************************************/
+/***************************** Include Files **********************************/
+/******************************************************************************/
+#include <drivers/tmr/adi_tmr.h>
+#include <drivers/pwr/adi_pwr.h>
+#include <stdlib.h>
+#include <math.h>
+#include "adc_update_timer.h"
+
+/******************************************************************************/
+/************************ Variables Definitions *******************************/
+/******************************************************************************/
+
+/* Update timer interrupt flag */
+uint8_t adc_channel_flag = 0;
+/* Software prescaler for the timer to extend the maximum time between
+ * interrupts */
+uint8_t adc_sw_prescaler = 0;
+
+/******************************************************************************/
+/************************ Functions Definitions *******************************/
+/******************************************************************************/
+
+/**
+ * Update timer interrupt callback function.
+ *
+ * Asserts the interrupt flag and increments the software prescaler. Standard
+ * format for callback functions.
+ *
+ * @param [in] pCBParam - Pointer to callback parameter list.
+ * @param [in] nEvent   - Interrupt source identifier.
+ * @param [in/out] pArg - Random argument list.(chosen by the application)
+ *
+ * @return none
+ */
+static void timer0_adc_update_callback(void *cb_param, uint32_t event,
+		void *arg)
+{
+	adc_channel_flag = 1;
+	adc_sw_prescaler++;
+}
+
+/**
+ * Calculate timer load value and clock prescaler based on the desired
+ * frequency.
+ *
+ * If the desired frequency is too low, use a software prescaler with maximum
+ * value of 32.
+ *
+ * @param [in/out] dev            - Update timer descriptor structure. Contains
+ *        the desired frequency as input parameter and the software prescaler as
+ *        output parameter.
+ * @param [out] ptr_tmr_load_val  - New load value of the timer.
+ * @param [out] ptr_tmr_prescaler - New timer prescaler value.
+ *
+ * @return 0 in case of success, negative error code otherwise.
+ */
+static int32_t adc_update_timer_get_config(struct adc_update_desc *dev,
+		uint16_t *ptr_tmr_load_val, ADI_TMR_PRESCALER *ptr_tmr_prescaler)
+{
+	uint32_t pclk_val;
+	float sw_divided_freq, prescaled_freq;
+	uint8_t i, j;
+	uint8_t done_flag;
+	uint32_t ret;
+
+	for(j = 1; j < 33; j++) {
```

Comment:
curious: why 33?

Reply:
Sorry for the magic number. Will replace it with a define. It's 33 because at first this clock division was done by a hardware prescaler that goes up to 32. But then I realized that if I prescale the clock for the peripherals I have to also prescale the clock for the core because the core clock must never be higher that the peripheral clock for some reason. So I just moved to a software prescaling instead and kept the 32 limit because I figured we will not need more. But it can be anything we want.